libraries/heft-config-file/src/ConfigurationFile.ts

```
@@ -300,36 +286,61 @@ export class ConfigurationFile<TConfigurationFile> {
-          resolvedConfigurationFilePath
-        )}" in ConfigurationFile path.`
+    let cacheEntryPromise: Promise<IConfigurationFileCacheEntry<TConfigurationFile>> | undefined =
+      this._configPromiseCache.get(resolvedConfigurationFilePath);
+    if (!cacheEntryPromise) {
+      cacheEntryPromise = this._createCacheEntryPromise(
+        this._loadConfigurationFileInnerAsync(
+          terminal,
+          resolvedConfigurationFilePath,
+          visitedConfigurationFilePaths,
+          rigConfig
+        )
       );
+      this._configPromiseCache.set(resolvedConfigurationFilePath, cacheEntryPromise);
     }
 
+    // We check for loops after caching a promise for this config file, but before attempting
+    // to resolve the promise. We can't handle loop detection in the `InnerAsync` function, because
+    // we could end up waiting for a cached promise (like A -> B -> A) that never resolves.
+    if (visitedConfigurationFilePaths.has(resolvedConfigurationFilePath)) {
+      const resolvedConfigurationFilePathForLogging: string = ConfigurationFile._formatPathForLogging(
+        resolvedConfigurationFilePath
+      );
+      throw new Error(
+        'A loop has been detected in the "extends" properties of configuration file at ' +
+          `"${resolvedConfigurationFilePathForLogging}".`
+      );
+    }
+    visitedConfigurationFilePaths.add(resolvedConfigurationFilePath);
+
+    const cacheEntry: IConfigurationFileCacheEntry<TConfigurationFile> = await cacheEntryPromise;
     if (cacheEntry.error) {
       throw cacheEntry.error;
     } else {
       return cacheEntry.configurationFile! as TConfigurationFile;
     }
   }
 
+  // Convert a promise for a TConfigurationFile into a promise for a corresponding
+  // cache entry instead.
+  private async _createCacheEntryPromise(
+    configFilePromise: Promise<TConfigurationFile>
+  ): Promise<IConfigurationFileCacheEntry<TConfigurationFile>> {
+    try {
```

You should just be able to do:
```
configFilePromise.catch(() => { /* suppress unhandled rejection errors */ });
return configFilePromise;
```

---

I don't think that would have the same API though.

This function converts one type of promise into a different type of promise:

 - The _input_ promise is a promise for a `TConfigurationFile`, and it will resolve if it exists and reject with an error if it does not.
 - The _output_ promise is a promise for a `IConfigurationFileCacheEntry`.  It should *resolve* with the object `{configurationFile: [thing]}` if the input promise resolves, and it should *resolve* with the object `{error: [thing]}` if the input promise rejects.

I don't think I can suppress any errors here, because I need to save that error in the cache entry.

---

I meant to imply changing the promise type to `TConfigurationFile`, since a rejected promise can be reused to save the error just as easily as caching it in an object.

---

Ohhh! Remove `Promise<IConfigurationFileCacheEntry>` completely and replace with `Promise< TConfigurationFile>`!

Makes total sense, I'll try that out and push it up if I don't run into any issues.